clang/lib/CIR/CodeGen/CIRAsm.cpp

```diff
@@ -524,15 +524,12 @@ mlir::LogicalResult CIRGenFunction::emitAsmStmt(const AsmStmt &S) {
   // If this is a Microsoft-style asm blob, store the return registers (EAX:EDX)
   // to the return value slot. Only do this when returning in registers.
   if (isa<MSAsmStmt>(&S)) {
-    const cir::ABIArgInfo &RetAI = CurFnInfo->getReturnInfo();
-    if (RetAI.isDirect() || RetAI.isExtend()) {
-      // Make a fake lvalue for the return value slot.
-      LValue ReturnSlot = makeAddrLValue(ReturnValue, FnRetTy);
-      CGM.getTargetCIRGenInfo().addReturnRegisterOutputs(
-          *this, ReturnSlot, Constraints, ResultRegTypes, ResultTruncRegTypes,
-          ResultRegDests, AsmString, S.getNumOutputs());
-      SawAsmBlock = true;
-    }
+    // Make a fake lvalue for the return value slot.
+    LValue ReturnSlot = makeAddrLValue(ReturnValue, FnRetTy);
```

Comment:
I wonder if all of this should also be moved to a lowering pass eventually. Doesn't this break the Windows ABI when we *don't* return in registers?

Reply:
You may be right. In particular, this is probably wrong for void functions. I was applying a mental rule "everything acts like "direct" but void returns act like "ignore" so I'll need to do something different here.

Reply:
More info to share here.... We don't have any tests currently for MS inline asm. I did verify that if I try to compile code with MS inline asm in a void function, it crashes here. Digging a bit deeper, I see that in classic codegen the `addReturnRegisterOutputs()` function does nothing by default and is only implemented for 32-bit x86 targets. If I compile for a 32-bit x86 target with CIR, it crashes because we don't have any target ABI support for that (and I don't think anyone is in a hurry to add it). I don't think we need to do anything here. 

I'd prefer not to add an ms-asm test now, because I don't want to spend the time verifying that the outputs we're producing now are correct. That seems better left until such a time as someone is actually working on inline asm.